hazelcast/include/hazelcast/client/serialization/pimpl/compact/compact_impl.h

```diff
@@ -574,18 +570,59 @@ struct schema_of
 template<typename T>
 const schema schema_of<T>::schema_v = schema_of<T>::build_schema();
 
+template<typename T>
+schema build_schema(const T& object)
+{
+    schema_writer schema_writer(hz_serializer<T>::type_name());
+    serialization::compact_writer writer =
+        create_compact_writer(&schema_writer);
+    serialization::hz_serializer<T>::write(object, writer);
+    return std::move(schema_writer).build();
+}
+
+template<typename T>
+class class_to_schema
+{
+public:
+
+    static const boost::optional<schema>& get()
+    {
+        std::lock_guard<std::mutex> guard{ mtx_ };
+
+        return value_;
+    }
+
+    static void set(const boost::optional<schema>& schema)
+    {
+        std::lock_guard<std::mutex> guard { mtx_ };
+
+        value_ = schema;
+    }
+
+private:
+
+    static std::mutex mtx_;
+    static boost::optional<schema> value_;
+};
+
+template<typename T>
+boost::optional<schema> class_to_schema<T>::value_ = boost::none;
+
+template<typename T>
+std::mutex class_to_schema<T>::mtx_;
+
 template<typename T>
 T inline compact_stream_serializer::read(object_data_input& in)
 {
     int64_t schema_id = in.read<int64_t>();
-    const auto& local_schema = schema_of<T>::schema_v;
-    // optimization to avoid hitting shared map in the schema_service,
-    // in the case incoming data's schema is same as the local schema
-    if (schema_id == local_schema.schema_id()) {
-        compact_reader reader = create_compact_reader(*this, in, local_schema);
+
+    const boost::optional<schema>& local_schema = class_to_schema<T>::get();
```

Comment:
same issue in PR #1039. Just want to mention again for merging.

Reply:
Fixed.